server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java

```diff
@@ -10,56 +10,155 @@
+        logger.debug("Received {} parts for blob {}", readContext.getNumberOfParts(), blobName);
         final int numParts = readContext.getNumberOfParts();
-        final AtomicBoolean anyPartStreamFailed = new AtomicBoolean();
-        FileCompletionListener fileCompletionListener = new FileCompletionListener(numParts, fileName, completionListener);
-
-        for (int partNumber = 0; partNumber < numParts; partNumber++) {
-            FilePartWriter filePartWriter = new FilePartWriter(
-                partNumber,
-                readContext.getPartStreams().get(partNumber),
-                fileLocation,
-                anyPartStreamFailed,
-                fileCompletionListener
-            );
-            threadPool.executor(ThreadPool.Names.GENERIC).submit(filePartWriter);
+        final AtomicBoolean anyPartStreamFailed = new AtomicBoolean(false);
+        final GroupedActionListener<String> groupedListener = new GroupedActionListener<>(
+            ActionListener.wrap(r -> completionListener.onResponse(blobName), completionListener::onFailure),
+            numParts
+        );
+        final Queue<ReadContext.StreamPartCreator> queue = new ConcurrentLinkedQueue<>(readContext.getPartStreams());
+        final StreamPartProcessor processor = new StreamPartProcessor(
+            queue,
+            anyPartStreamFailed,
+            fileLocation,
+            groupedListener,
+            threadPool.executor(ThreadPool.Names.REMOTE_RECOVERY),
+            rateLimiter
+        );
+        // We have a dedicated thread pool for draining streams and writing them to a file on the
+        // local disk. However, we want more concurrent requests to fetch stream parts to be
+        // happening asynchronously so that there is always a queue of streams ready to be consumed
+        // by the REMOTE_RECOVERY thread pool.
+        final int concurrentStreamRequests = 20;
+        for (int i = 0; i < Math.min(concurrentStreamRequests, queue.size()); i++) {
+            processor.process(queue.poll());
         }
     }
 
     @Override
     public void onFailure(Exception e) {
         completionListener.onFailure(e);
     }
+
+    private static class StreamPartProcessor {
+        private static final RuntimeException CANCELED_PART_EXCEPTION = new RuntimeException(
+            "Canceled part download due to previous failure"
+        );
+        private final Queue<ReadContext.StreamPartCreator> queue;
+        private final AtomicBoolean anyPartStreamFailed;
+        private final Path fileLocation;
+        private final GroupedActionListener<String> completionListener;
+        private final Executor executor;
+        private final UnaryOperator<InputStream> rateLimiter;
+
+        private StreamPartProcessor(
+            Queue<ReadContext.StreamPartCreator> queue,
+            AtomicBoolean anyPartStreamFailed,
+            Path fileLocation,
+            GroupedActionListener<String> completionListener,
```

Comment:
```suggestion
            ActionListener<String> completionListener,
```

Reply:
This was intentional. The semantics of a GroupedActionListener requires that it be notified an exact number of times (success or failure). There's a fair bit of logic here to accomplish that so I thought the specific type helped convey that intent.